onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,709 @@
+      FuseNodes(graph, output, cast_fp16_siblings, removed_nodes);
+      VLOGS(logger, 1) << "FusedSubgraphs: Fused Cast nodes : " << ConcatNames<std::vector<Node*>>(cast_fp16_siblings);
+    }
+    if (cast_fp32_siblings.size() > 1) {
+      modified = true;
+      FuseNodes(graph, output, cast_fp32_siblings, removed_nodes);
+      VLOGS(logger, 1) << "FusedSubgraphs: Fused Cast nodes : " << ConcatNames<std::vector<Node*>>(cast_fp32_siblings);
+    }
+  }
+  return modified;
+}
+
+// RemoveUnnecessaryCasts
+// Remove a cast if the input elem_type is same the required cast type.
+static bool RemoveUnnecessaryCasts(Graph& graph, Node* node,
+                                   std::deque<onnxruntime::NodeIndex>& removed_nodes,
+                                   const logging::Logger& logger) {
+  bool modified = false;
+  if (node->InputDefs().size() == 1) {
+    const NodeArg* node_arg = node->InputDefs()[0];
+    auto elem_type = node_arg->TypeAsProto()->tensor_type().elem_type();
+    TensorProto_DataType data_type = static_cast<TensorProto_DataType>(elem_type);
+    if (IsCastTo(node, data_type)) {
+      VLOGS(logger, 1) << "Removed unnecessary cast " << node->Name();
+      RemoveCastNodes(graph, {node}, removed_nodes);
+      modified = true;
+    }
+  }
+  return modified;
+}
+
+// PropagateFP32CastsFromInputsToOutputs
+// This non recursive fusion, checks whether the given node is fp16 safe op and
+// whether all floatingpoint inputs are cast to fp32
+// and propagates cast op to the floatingpoint outputs.
+static bool PropagateFP32CastsFromInputsToOutputs(Graph& graph, Node* node,
+                                                  std::deque<onnxruntime::NodeIndex>& removed_nodes,
+                                                  size_t level,
+                                                  const logging::Logger& logger) {
+  bool modified = false;
+  if (IsFP16Allow(node->OpType(), level)) {
+    bool has_float_inputs = false;
+    bool all_float_inputs_have_casts = true;
+    std::vector<Node*> casts;
+    std::unordered_set<NodeArg*> require_type_change;
+    for (NodeArg* input : node->MutableInputDefs()) {
+      if (!IsType(*input, TensorProto::FLOAT)) {
+        continue;
+      }
+      has_float_inputs = true;
+      Node* producer = graph.GetMutableProducerNode(input->Name());
+      if (producer && IsCastTo(producer, TensorProto::FLOAT)) {
+        casts.push_back(producer);
+        require_type_change.insert(input);
+        continue;
+      }
+      all_float_inputs_have_casts = false;
+      break;
+    }
+    if (has_float_inputs && all_float_inputs_have_casts && casts.size() > 1) {
+      VLOGS(logger, 1) << "PropagateFP32CastsFromInputsToOutputs: Removed Cast nodes "
+                       << ConcatNames<std::vector<Node*>>(casts)
+                       << " feeding the same compute node " << node->Name();
+      for (Node* cast : casts) {
+        RemoveCastNodes(graph, {cast}, removed_nodes);
+      }
```

Comment:
My first reaction is "Why we can't pass all the casts in one shot"... latter I realized that we RemoveCastNodes() is acctually assuming the input casts to be sequential... 

In this case, maybe consider using graph_utils::RemoveNode ?

Reply:
RemoveNode is delayed to the end of the loop. We want to update the graph nodes and edges for the next functions.
Instead of implementing two separate functions, RemoveCastNode and RemoveCastNodes, I implemented just second one that take a chain of cast nodes, and calling it with one element when I need to remove a single nodes. This works as expected.